synthesizer/src/process/trace/inclusion/mod.rs

```diff
@@ -91,44 +94,62 @@ impl<N: Network> Inclusion<N> {
-        // Initialize the global state root.
-        let mut global_state_root = N::StateRoot::default();
-        // Initialize a vector for the batch assignments.
-        let mut batch_assignments = vec![];
-
-        for assignment in assignments.iter() {
-            // Ensure the global state root is the same across iterations.
-            if *global_state_root != Field::zero() && global_state_root != assignment.state_path.global_state_root() {
-                bail!("Inclusion expected the global state root to be the same across iterations")
+impl<N: Network> Inclusion<N> {
+    /// Returns the verifier public inputs for the given global state root and transitions.
+    pub fn prepare_verifier_inputs<'a>(
+        global_state_root: N::StateRoot,
+        transitions: impl ExactSizeIterator<Item = &'a Transition<N>>,
+    ) -> Result<Vec<Vec<N::Field>>> {
+        // Determine the number of transitions.
+        let num_transitions = transitions.len();
+
+        // Initialize an empty transaction tree.
+        let mut transaction_tree = N::merkle_tree_bhp::<TRANSACTION_DEPTH>(&[])?;
+        // Initialize a vector for the batch verifier inputs.
+        let mut batch_verifier_inputs = vec![];
+
+        // Construct the batch verifier inputs.
+        for (transition_index, transition) in transitions.enumerate() {
+            // Retrieve the local state root.
+            let local_state_root = *transaction_tree.root();
+
+            // Iterate through the inputs.
+            for input in transition.inputs() {
+                // Filter the inputs for records.
+                if let Input::Record(serial_number, _) = input {
+                    // Add the public inputs to the batch verifier inputs.
+                    batch_verifier_inputs.push(vec![
+                        N::Field::one(),
+                        **global_state_root,
+                        *local_state_root,
+                        **serial_number,
+                    ]);
+                }
             }
-            // Update the global state root.
-            global_state_root = assignment.state_path.global_state_root();
 
-            // Add the assignment to the assignments.
-            batch_assignments.push(assignment.to_circuit_assignment::<A>()?);
+            // If this is not the last transition, append the transaction leaf to the transaction tree.
+            if transition_index + 1 != num_transitions {
+                // Construct the transaction leaf.
+                let transaction_leaf = TransactionLeaf::new_execution(transition_index as u16, **transition.id());
+                // Insert the leaf into the transaction tree.
+                transaction_tree.append(&[transaction_leaf.to_bits_le()])?;
+            }
         }
 
         // Ensure the global state root is not zero.
-        if *global_state_root == Field::zero() {
+        if batch_verifier_inputs.is_empty() && *global_state_root == Field::zero() {
```

Comment:
Why do we only want a non-zero `global_state_root` when `batch_verifier_inputs.is_empty()`?

Reply:
To preserve privacy in the case that we introduce function privacy in the future.